ranger/ext/img_display.py

```diff
@@ -67,6 +68,47 @@ def move_cur(to_y, to_x):
     bin_stdout.write(tparm)
 
 
+def get_terminal_size():
+    farg = struct.pack("HHHH", 0, 0, 0, 0)
+    fd_stdout = sys.stdout.fileno()
+    fretint = fcntl.ioctl(fd_stdout, termios.TIOCGWINSZ, farg)
+    return struct.unpack("HHHH", fretint)
+
+
+def get_font_dimensions():
+    """
+    Get the height and width of a character displayed in the terminal in
+    pixels.
+    """
+    rows, cols, xpixels, ypixels = get_terminal_size()
+    return (xpixels // cols), (ypixels // rows)
+
+
+def image_fit_width(width, height, max_cols, max_rows, font_width=None, font_height=None):
+    if font_width is None or font_height is None:
+        font_width, font_height = get_font_dimensions()
+
+    max_width = font_width * max_cols
+    max_height = font_height * max_rows
+    if height > max_height:
+        if width > max_width:
+            width_scale = max_width / width
+            height_scale = max_height / height
+            min_scale = min(width_scale, height_scale)
+            max_scale = max(width_scale, height_scale)
+            if width * max_scale <= max_width and height * max_scale <= max_height:
```

Comment:
👍 on refactoring this into a common function.

I'm having trouble figuring out if the logic actually makes sense though. Paraphrased:
```
height < max_height and width < max_width -> don't scale
height < max_height and width > max_width -> scale to fit width
height > max_height and width < max_width -> scale width so height fits
height > max_height and width > max_width -> scale so the entire image fits
```
Surely in that last case we should always scale with `min_scale`? Checking both scales only makes sense to me if we don't know whether both dimensions exceed the maximum.

Reply:
So scaling is likely the point that impacts performance the most. Either using multi threading or GPU acceleration could noticeably impact usability (not in this PR, but long term).

Reply:
I'd still appreciate an answer to this question. When could the `max_scale` ever work?